File: codigo/python/simpson.py
"""simpson.py -- Integracion por la regla de Simpson 1/3 compuesta. Chapra cap. 21."""


def simpson(f, a, b, n=100):
    if n % 2 == 1:                 # Simpson 1/3 requiere n par
        n += 1
    h = (b - a) / n
    s = f(a) + f(b)
    for i in range(1, n):
        s += (4 if i % 2 else 2) * f(a + i * h)
    return s * h / 3


if __name__ == "__main__":
    import math
    I = simpson(math.sin, 0, math.pi, 100)
    print(f"Integral de sin en [0, pi] = {I:.10f}   (exacto = 2)")

File: codigo/matlab/simpson.m
function I = simpson(f, a, b, n)
% SIMPSON  Regla de Simpson 1/3 compuesta (n par). Chapra cap. 21.
%   Uso:  I = simpson(@sin, 0, pi, 100);
    if nargin < 4, n = 100; end
    if mod(n,2) == 1, n = n + 1; end          % n debe ser par
    h = (b - a)/n;
    x = a:h:b;
    y = f(x);
    I = (h/3)*(y(1) + 4*sum(y(2:2:end-1)) + 2*sum(y(3:2:end-2)) + y(end));
end
